Duplicati/Library/Main/Operation/DeleteHandler.cs

```diff
@@ -204,11 +204,18 @@ private DateTime[] GetFilesetsToDelete(Database.LocalDeleteDatabase db, DateTime
                 }
             }
 
-            // Remove backups that are older than date specified via option
+            // Remove backups that are older than date specified via option while ensuring
+            // that we always have at least one full backup.
             var keepTime = m_options.KeepTime;
             if (keepTime.Ticks > 0)
             {
-                toDelete.AddRange(sortedAllBackups.SkipWhile(x => x >= keepTime));
+                bool haveFullBackup = false;
+                toDelete.AddRange(sortedAllBackups.SkipWhile(x =>

```

Comment:
What about 
                toDelete.AddRange(sortedAllBackups.Where(x => x <= keepTime)
                    .Where(x =>
                    {
                        haveFullBackup = haveFullBackup || db.IsFilesetFullBackup(x);
                        return haveFullBackup;
                    }));
?

Reply:
Since `sortedAllBackups` are ordered (descending), I liked how `SkipWhile` would avoid enumerating the entire sequence.

Also, I'm not sure if I'm reading it correctly, but does your example only look for "old" backups that are full?  If one of the "recent" backups is full, then there's no need to keep any backups prior to `keepTime`.

Reply:
That's true, then your code is correct. I had to think about your code yesterday evening for a second, I understood it the way you explained. Could you write a test for it that we make sure that logic is preserved?
(btw. I need more time to review this since this PR is a bit more delicate)

Reply:
> I need more time to review this since this PR is a bit more delicate

No problem.  Thanks for your help reviewing.  In case you haven't noticed, there is a lot of related discussion in issue #3982 that may help explain some of the changes made here.

Reply:
Revision afecf891dab745a7f06e74e9b47687e12a09e3a1 adds a test for the `--keep-time` retention logic.  I'll try and write a test for the `--keep-versions` logic soon.